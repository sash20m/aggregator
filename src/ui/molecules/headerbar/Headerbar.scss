.header {
    display: flex;
    width: 100%;
    background-color: rgb(255, 255, 255);
    border-bottom: 1px solid rgb(223, 223, 223);
    align-items: center;
}

.logo {
    margin: 15px;

    &:hover {
        cursor: pointer;
    }
}

.search {
    margin-left: 50px;

    &__input {
        width: 50vw;
        height: 4vh;
        border: 1px solid #9aabc2;
        color: rgba(0, 0, 0, 0.65);
        font-size: 1.3em;
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
        line-height: 1.5;
        border-radius: 20px;
        padding-left: 30px;
        outline: none;

        &--active {
            width: 50vw;
            height: 4vh;
            border: 1px solid #467bbc;
            color: rgba(0, 0, 0, 0.65);
            font-size: 1.3em;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
            line-height: 1.5;
            padding-left: 30px;
            outline: none;
            border-radius: 31px 31px 0 0;
        }
    }

    &__suggestions {
        position: absolute;
        z-index: 2;
        width: 50.15vw;
        overflow-y: auto;
        list-style: none;
        padding-left: 0;
        background: white;
        border-top-width: 0;
        border-bottom-left-radius: 20px;
        border-bottom-right-radius: 20px;
        border-left: 1px solid #467bbc;
        border-right: 1px solid #467bbc;
        border-bottom: 1px solid #467bbc;
        padding-right: 30px;
        margin-top: 0;

        &__item {
            position: relative;
            overflow: hidden;
            font-size: 1em;
            font-weight: 600;
            line-height: 22px;
            color: #000;
            white-space: nowrap;
            text-overflow: ellipsis;
            cursor: pointer;
            box-sizing: border-box;
            list-style: none;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;

            &__button {
                width: 100%;
                text-align: left;
                padding: 10px 25px;
                border: none;
                margin: 0;
                overflow: visible;
                background: transparent;
                color: inherit;
                font: inherit;
                line-height: normal;
                -webkit-font-smoothing: inherit;
                -moz-osx-font-smoothing: inherit;
                -webkit-appearance: none;
                cursor: pointer;
                outline: inherit;

                &:hover {
                    color: #056ac8;
                }
            }
        }
    }

    &__options {
        display: flex;
        flex-direction: row;
        align-items: center;
        margin-top: 20px;

        &:hover {
            cursor: pointer;
        }

        &__item {
            font-size: 1.1em;
            color: rgb(61, 61, 61);
            font-weight: 500;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
            margin: 5px;

            &--selected {
                font-size: 0.9em;
                color: #467bbd;
                font-weight: 500;
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
                margin: 5px;
                border-radius: 20px;
                border: 1px solid rgb(121, 121, 121);
                padding: 10px 15px 10px 15px;
                margin-left: 10px;
                margin-right: 10px;
            }
        }
    }
}