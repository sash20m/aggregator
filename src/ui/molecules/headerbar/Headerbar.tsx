/* eslint-disable jsx-a11y/mouse-events-have-key-events */
import React, { useState, useEffect } from 'react';

import Link from 'next/link';
import { useRouter } from 'next/router';
import { getSuggestions } from 'services/companies.services';
import { Suggestions } from 'index';

import './Headerbar.scss';

interface Props {
  searchVisible: boolean;
}

export const HeaderBar = ({ searchVisible }: Props): JSX.Element => {
  const Router = useRouter();
  const [inputValue, setInputValue] = useState<string>('');
  const [focused, setFocused] = useState<boolean>(false);
  const [mouseOver, setMouseOver] = useState<boolean>(false);
  const [data, setData] = useState<Suggestions[] | null>(null);

  const getSuggestionsList = async (name: string | null) => {
    try {
      const res = await getSuggestions(name);
      setData(res);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
    }
  };

  useEffect(() => {
    getSuggestionsList(null);
  }, []);

  const searchCompanies = (e: React.ChangeEvent<HTMLInputElement>) => {
    getSuggestionsList(e.target.value);
    setInputValue(e.target.value);
  };

  const onInputClick = () => {
    setFocused(true);
  };

  const onInputLeave = () => {
    if (!mouseOver) setFocused(false);
  };

  const mouseOverList = () => {
    setMouseOver(true);
  };

  const mouseLeaveList = () => {
    setMouseOver(false);
  };

  const goToCompanies = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      sessionStorage.setItem('lastSearched', inputValue);
      Router.push(`/companies/${inputValue}`);
    }
  };

  const resetSessionStorage = () => {
    sessionStorage.clear();
  };

  const showAllCompanies = () => {
    setFocused(false);
    Router.push('/companies/all');
  };

  const resetSuggestionList = () => {
    setFocused(false);
  };

  return (
    <div className="header-main" data-testid="headerbar">
      <div className="header-main__content">
        <Link href="/">
          <button
            className="logo__btn"
            type="button"
            onClick={resetSessionStorage}
          >
            <img className="logo" src="../logo-small.png" alt="Logo" />
          </button>
        </Link>
        {searchVisible && (
          <div className="search">
            <img className="search__icon" src="../search.png" alt="" />
            <input
              className={focused ? 'search__input--active' : 'search__input'}
              type="text"
              placeholder="Search from 225,195 companies"
              onClick={onInputClick}
              onBlur={onInputLeave}
              onChange={searchCompanies}
              value={inputValue}
              onKeyDown={goToCompanies}
            />
            <button
              type="button"
              className="search__show-all"
              onClick={showAllCompanies}
            >
              Show All
            </button>

            {focused && (
              <ul
                className="search__suggestions"
                onMouseOver={mouseOverList}
                onMouseOut={mouseLeaveList}
              >
                {!data && (
                  <p className="search__suggestions__loading"> Loading...</p>
                )}

                {data &&
                  data.map((company: Suggestions) => (
                    <Link
                      href="/company/[slug]"
                      as={`/company/${company.slug}`}
                    >
                      <li
                        key={company.name}
                        className="search__suggestions__item"
                      >
                        <button
                          type="button"
                          className="search__suggestions__item__button"
                          name={company.slug}
                          onClick={resetSuggestionList}
                        >
                          {company.name}
                        </button>
                      </li>
                    </Link>
                  ))}
              </ul>
            )}
          </div>
        )}
      </div>
    </div>
  );
};
