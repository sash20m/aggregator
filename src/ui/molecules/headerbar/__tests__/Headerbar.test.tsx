import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';
import { HeaderBar } from '../Headerbar';

afterEach(cleanup);

const props = {
  searchVisible: false,
};

describe('headerbar', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<HeaderBar searchVisible />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render component', () => {
    const { queryByTestId } = render(
      <HeaderBar searchVisible={props.searchVisible} />
    );

    expect(queryByTestId('headerbar')).toBeInTheDocument();
  });

  it('it should render search', () => {
    const { queryByPlaceholderText } = render(
      <HeaderBar searchVisible={props.searchVisible} />
    );

    const element = queryByPlaceholderText('Search from 225,195 companies');
    const t = expect(element);

    if (props.searchVisible) t.toBeInTheDocument();
    else t.toBeFalsy();
  });
});
