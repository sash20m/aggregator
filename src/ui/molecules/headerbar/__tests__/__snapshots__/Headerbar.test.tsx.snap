// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headerbar should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="header-main"
    data-testid="headerbar"
  >
    <div
      class="header-main__content"
    >
      <button
        class="logo__btn"
        type="button"
      >
        <img
          alt="Logo"
          class="logo"
          src="../logo-small.png"
        />
      </button>
      <div
        class="search"
      >
        <img
          alt=""
          class="search__icon"
          src="../search.png"
        />
        <input
          class="search__input"
          placeholder="Search from 225,195 companies"
          type="text"
          value=""
        />
        <button
          class="search__show-all"
          type="button"
        >
          Show All
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
